Source/Core/Core/AchievementManager.cpp

```diff
@@ -152,66 +158,218 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
-    if (start_session_response != ResponseType::SUCCESS)
+    m_loading_volume = DiscIO::CreateVolume(volume->GetBlobReader().CopyReader());
+  }
+  m_system = &Core::System::GetInstance();
+  struct FilereaderState
+  {
+    int64_t position = 0;
+    std::unique_ptr<DiscIO::Volume> volume;
+  };
+  rc_hash_filereader volume_reader{
+      .open = [](const char* path_utf8) -> void* {
+        auto state = std::make_unique<FilereaderState>();
+        {
+          std::lock_guard lg{*AchievementManager::GetInstance()->GetLock()};
+          state->volume = std::move(AchievementManager::GetInstance()->GetLoadingVolume());
+        }
+        if (!state->volume)
+          return nullptr;
+        return state.release();
+      },
+      .seek =
+          [](void* file_handle, int64_t offset, int origin) {
+            switch (origin)
+            {
+            case SEEK_SET:
+              reinterpret_cast<FilereaderState*>(file_handle)->position = offset;
+              break;
+            case SEEK_CUR:
+              reinterpret_cast<FilereaderState*>(file_handle)->position += offset;
+              break;
+            case SEEK_END:
+              // Unused
+              break;
+            }
+          },
+      .tell =
+          [](void* file_handle) {
+            return reinterpret_cast<FilereaderState*>(file_handle)->position;
+          },
+      .read =
+          [](void* file_handle, void* buffer, size_t requested_bytes) {
+            FilereaderState* filereader_state = reinterpret_cast<FilereaderState*>(file_handle);
+            bool success = (filereader_state->volume->Read(
+                filereader_state->position, requested_bytes, reinterpret_cast<u8*>(buffer),
+                DiscIO::PARTITION_NONE));
+            if (success)
+            {
+              filereader_state->position += requested_bytes;
+              return requested_bytes;
+            }
+            else
+            {
+              return static_cast<size_t>(0);
+            }
+          },
+      .close = [](void* file_handle) { delete reinterpret_cast<FilereaderState*>(file_handle); }};
+  rc_hash_init_custom_filereader(&volume_reader);
+  m_queue.EmplaceItem([this, callback] {
+    Hash new_hash;
+    if (!rc_hash_generate_from_file(new_hash.data(), RC_CONSOLE_GAMECUBE, ""))
     {
-      callback(start_session_response);
-      WARN_LOG_FMT(ACHIEVEMENTS, "Failed to connect to RetroAchievements server.");
-      OSD::AddMessage("Failed to connect to RetroAchievements server.", OSD::Duration::VERY_LONG,
-                      OSD::Color::RED);
+      ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from volume.");
+      callback(AchievementManager::ResponseType::MALFORMED_OBJECT);
       return;
     }
-
-    const auto fetch_game_data_response = FetchGameData();
-    if (fetch_game_data_response != ResponseType::SUCCESS)
+    if (AchievementManager::GetInstance()->IsDisabled())
```

Comment:
This should probably be checked under the lock to avoid timing issues.

Reply:
Oh and you already have the `this` pointer here, no need to fetch it via `GetInstance()`.